pkg/stats/statser_internal.go

```diff
@@ -37,30 +33,26 @@ func NewInternalStatser(tags gostatsd.Tags, namespace string, hostname gostatsd.
 		tags = tags.Concat(gostatsd.Tags{"host:" + string(hostname)})
 	}
 	return &InternalStatser{
-		buffer:    make(chan *gostatsd.Metric, bufferSize),
 		tags:      tags,
 		namespace: namespace,
 		hostname:  hostname,
 		handler:   handler,
+		// We can't just use a MetricMap because everything
+		// that writes to it is on its own goroutine.
+		consolidator: gostatsd.NewMetricConsolidator(10, 0, nil),
 	}
 }
 
-// Run will pull internal metrics off a small buffer, and dispatch them.  It
-// stops running when the context is closed.
-func (is *InternalStatser) Run(ctx context.Context) {
-	flushed, unregister := is.RegisterFlush()
-	defer unregister()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case m := <-is.buffer:
-			is.dispatchMetric(ctx, m)
-		case <-flushed:
-			is.Gauge("internal_dropped", float64(atomic.LoadUint64(&is.dropped)), nil)
-		}
+func (is *InternalStatser) NotifyFlush(ctx context.Context, d time.Duration) {
+	mms := is.consolidator.Drain(ctx)
+	if mms == nil {
+		// context is canceled
```

Comment:
You would only know that for sure if you checked `ctx.Error()` right?

Reply:
I'm breaking the abstraction and saying "yeah, I know why this failed". It's part of my DRAEASWH system.